runtime/runtime/src/verifier.rs

```diff
@@ -20,10 +22,98 @@ use near_store::{
 };
 
 use crate::config::{total_prepaid_gas, tx_cost, TransactionCost};
+use crate::near_primitives::account::{Account, FunctionCallPermission};
+use crate::near_primitives::trie_key::trie_key_parsers;
 use crate::VerificationResult;
-use near_primitives::checked_feature;
-use near_primitives::runtime::config::RuntimeConfig;
-use near_primitives::types::BlockHeight;
+use near_primitives::hash::CryptoHash;
+
+/// Checks if given account has enough balance for storage stake, and returns:
+///  - None if account has enough balance or is a zero-balance account
+///  - Some(insufficient_balance) if account doesn't have enough and how much need to be added,
+///  - Err(message) if account has invalid storage usage or amount/locked.
+///
+/// Read details of state staking
+/// <https://nomicon.io/Economics/README.html#state-stake>.
+pub fn get_insufficient_storage_stake(
+    account_id: &AccountId,
+    account: &Account,
+    runtime_config: &RuntimeConfig,
+    state_update: &mut TrieUpdate,
+    current_protocol_version: ProtocolVersion,
+) -> Result<Option<Balance>, String> {
+    let required_amount = Balance::from(account.storage_usage())
+        .checked_mul(runtime_config.storage_amount_per_byte())
+        .ok_or_else(|| {
+            format!("Account's storage_usage {} overflows multiplication", account.storage_usage())
+        })?;
+    let available_amount = account.amount().checked_add(account.locked()).ok_or_else(|| {
+        format!(
+            "Account's amount {} and locked {} overflow addition",
+            account.amount(),
+            account.locked()
+        )
+    })?;
+    if available_amount >= required_amount {
+        Ok(None)
+    } else {
+        // Check if the account is a zero balance account. The check is delayed until here because
+        // it requires storage reads
+        if checked_feature!(
+            "protocol_feature_zero_balance_account",
+            ZeroBalanceAccount,
+            current_protocol_version
+        ) && is_zero_balance_account(account_id, account, state_update)
+            .map_err(|e| e.to_string())?
+        {
+            return Ok(None);
+        }
+        Ok(Some(required_amount - available_amount))
+    }
+}
+
+fn is_zero_balance_account(
```

Comment:
add a comment that this is for NEP448 - and the exact conditions when account is zero balance are described there